runtime/download.go

```diff
@@ -0,0 +1,120 @@
+package runtime
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/user"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/yourbase/yb/plumbing/log"
+)
+
+const defaultCacheDir = "/tmp/yourbase/cache"
+
+func localCacheDir() string {
+	if cacheDir, exists := os.LookupEnv("YB_CACHE_DIR"); exists {
+		return cacheDir
+	}
+	// Resolve current homeDir
+	if u, err := user.Current(); err == nil && u.HomeDir != "" {
+		return filepath.Join(u.HomeDir, ".cache", "yourbase")
+	}
+	// Try again
+	if homeDir, exists := os.LookupEnv("HOME"); exists {
+		return filepath.Join(homeDir, ".cache", "yourbase")
+	}
+	return defaultCacheDir
+}
+
+func cacheFilenameForURL(url string) (string, error) {
+	reg, err := regexp.Compile("[^a-zA-Z0-9.]+")
+	if err != nil {
+		return "", fmt.Errorf("Can't compile regex: %v", err)
+	}
+	fileName := reg.ReplaceAllString(url, "")
+	return fileName, nil
+}
+
+func downloadFileWithCache(ctx context.Context, url string) (string, error) {
+	cacheDir := localCacheDir()
+
+	filename, err := cacheFilenameForURL(url)
+	if err != nil {
+		return "", err
+	}
+
+	os.MkdirAll(cacheDir, 0700)
+
+	cacheFilename := filepath.Join(cacheDir, filename)
+	log.Infof("Downloading %s to cache as %s", url, cacheFilename)
+
+	fileExists := false
+	fileSizeMismatch := false
+
+	// Exists, don't re-download
+	if fi, err := os.Stat(cacheFilename); !os.IsNotExist(err) && fi != nil {
+		fileExists = true
+
+		// try HEAD'ing the URL and comparing to local file
+		resp, err := http.Head(url)
+		if err == nil {
+			// checks response HTTP status
+			// 404, 500 or others like it
+			if strings.HasPrefix(resp.Status, "40") || strings.HasPrefix(resp.Status, "50") {
```

Comment:
While this works, using `resp.StatusCode` is more robust.

```suggestion
			if resp.StatusCode != http.StatusOK {
```

Reply:
HTTP Statuses 20{1,2,6} etc, would be treated as an error, but this simplistic download manager isn't resuming downloads yet, so moving forward. Thanks!